### ISSUE TITLE
Hot reload not working when sidebar imports variables

### ISSUE THREAD

**Issue by dsalaza4** [facebook/docusaurus#5039]
### Prerequisites

- [ X] I'm using the latest version of Docusaurus.
- [ X] I have tried the `npm run clear` or `yarn clear` command.
- [ X] I have tried `rm -rf node_modules yarn.lock package-lock.json` and re-installing packages.
- [ X] I have tried creating a repro with https://new.docusaurus.io

### Description

In order to avoid having a huge `sidebar.js` file,
we created a `main.js` file and [imported sidebars for different sections from external js files](https://gitlab.com/fluidattacks/product/-/tree/cf751f2b62f08e529dd4a13e694406d03e8f1418/docs/src/sidebar).

We then [specified main.js as the sidebarPath](https://gitlab.com/fluidattacks/product/-/blob/cf751f2b62f08e529dd4a13e694406d03e8f1418/docs/src/docusaurus.config.js#L117).

The problem with this approach is:

1. Every time that an imported file changes, the devServer does not recognize the change, thus not updating the sidebar accordingly.
1. Even if the `main.js` file changes, it looks like it is caching the imported variables from the other files, which also prevents the update.

### Have you read the [Contributing Guidelines on issues](https://github.com/facebook/docusaurus/blob/master/CONTRIBUTING.md#reporting-new-issues)?

Yes I have.

## Steps to reproduce

1. Use a `main.js` file for your sidebar that imports sub-sidebars from other files like
[here](https://gitlab.com/fluidattacks/product/-/blob/cf751f2b62f08e529dd4a13e694406d03e8f1418/docs/src/sidebar/main.js).
1. Start the development server with `docusaurus start`.
1. Modify one of the imported files.
1. Notice how the dev server does not notice the change.

## Expected behavior

Docusaurus should be able to notice which files are being imported

## Actual behavior

(Write what happened. Add full console log messages and screenshots, if applicable.)

## Your environment

- Public source code:
- Public site URL:
- Docusaurus version used:
- Environment name and version (e.g. Chrome 78.0.3904.108, Node.js 10.17.0):
- Operating system and version (e.g. Ubuntu 20.04.2 LTS):

## Reproducible demo

Use https://new.docusaurus.io to create a CodeSandbox reproducible demo of the bug.

(Paste the link to an example repo, including a `docusaurus.config.js`, and exact instructions to reproduce the issue.)

**Comment by slorber** [comment]
Agree, the current code only watch for changes in the sidebar path to reload the docs plugin content.

We should walk the dependency graph and also watch the required files.

This lib looks useful but not very modern (does not support ESM, which we may need soon), https://github.com/browserify/module-deps

If you find anything better that we could use, let me know

**Comment by dsalaza4** [comment]
Would it be possible, as a first step, that instead of trying to automatically generate the graph, we could explicitly provide the files to the config?

That way we could trigger an update if one of the provided files changes.

What I still do not know how to fix is avoiding caching requires, as it does not really matter if the `main.js` files changes, it still uses cached versions of the imported variables :confused:

**Comment by dacevedo12** [comment]
There seems to be another problem beyond the file watcher. Even if chokidar triggers the reload, the content doesn't change due to `require` being cached.

Shouldn't WDS take care of that by default?

**Comment by slorber** [comment]
The community answered on twitter with a few interesting tools to get the depenend files of your sidebar js module: https://twitter.com/NicoloRibaudo/status/1407722571722858501

> Would it be possible, as a first step, that instead of trying to automatically generate the graph, we could explicitly provide the files to the config?

We don't have a very good API for that but you can "enhance" an existing plugin (see comments here: https://github.com/facebook/docusaurus/issues/4138#issuecomment-771042127)

You should be able to wrap the default docs `getPathsToWatch` lifecycle (https://docusaurus.io/docs/lifecycle-apis#getpathstowatch) to add the additional paths you want to the list.

> What I still do not know how to fix is avoiding caching requires, as it does not really matter if the main.js files changes, it still uses cached versions of the imported variables 😕

We use this to avoid caching issues inside Docusaurus: https://github.com/sindresorhus/import-fresh, maybe it will work for you too
